### ISSUE TITLE
32-bit PackedTensorAccessor

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#19268]
According to @t-vi and Clement Pinard, PackedTensorAccessor is slower than THCDeviceTensor because it does 64-bit indexing, and 32-bit indexing can be quite a bit faster on CUDA. We should support it as an option (with documentation about how to properly test if you can use 32-bit indexing or not.)

cc @ezyang @gchanan

**Comment by t-vi** [comment]
We do?

Am 15. April 2019 20:52:42 MESZ schrieb "Edward Z. Yang" <notifications@github.com>:

**Comment by ezyang** [comment]
You're right, egg on my face. So this is more of a documentation request.

**Comment by ezyang** [comment]
@gchanan I'm going to disagree with the priority label here. This is a very serious gotcha (and hard to see in code review) for use of PackedTensorAccessor which I have seen multiple people trip over over the past few months.

**Comment by gchanan** [comment]
@ezyang can you link a few examples?

**Comment by ezyang** [comment]
This one from Clement Pinard on Slack:

> Also has there been speed tests with packed accessors ? In my correlation module I pushed a commit ( https://github.com/ClementPinard/Pytorch-Correlation-extension/commit/157dae0e097a7f3f108777063f94a147d73beabf ) that used packed_accessor instead of THCDeviceTensor, and my forward pass went from ~5ms to ~7ms in my benchmarks (GTX 980 Ti)
The benchmark used is https://github.com/ClementPinard/Pytorch-Correlation-extension/blob/master/benchmark.py

Also this PR from Ivan https://github.com/pytorch/pytorch/pull/19630 ported PackedTensorAccessor directly, causing indexing to change from 32-bit to 64-bit. Probably every porting PR which had code previously using THCDeviceTensor also committed the same mistake.

**Comment by gchanan** [comment]
ok, you convinced me :).

**Comment by t-vi** [comment]
Batch norm saw a huge impact, too.

**Comment by t-vi** [comment]
So where are we aiming to have that? In the API docs, https://pytorch.org/cppdocs/api/classat_1_1_tensor.html?highlight=accessor#_CPPv4NK2at6Tensor8accessorEv , the tutorial https://pytorch.org/tutorials/advanced/cpp_extension.html#using-accessors (which is using `size_t` at the moment...)?

**Comment by gchanan** [comment]
@ezyang do you have a recommendation here?
